File: llm_harness/utils/youtube_utils.py
"""Shared YouTube text and URL utilities."""

from __future__ import annotations

import re

YOUTUBE_PATTERNS = (
    r"youtube\.com/watch\?v=",
    r"youtu\.be/",
    r"youtube\.com/embed/",
    r"youtube\.com/v/",
)


def clean_text(text: str) -> str:
    """Clean text by removing excessive whitespace and normalizing."""
    text = re.sub(r"\n{3,}", r"\n\n", text)
    text = re.sub(r" {2,}", " ", text)
    return text.strip()


def clean_youtube_url(url: str) -> str:
    """Normalize YouTube URL variants to watch URL when possible."""
    if "youtube.com/watch" in url:
        match = re.search(r"v=([a-zA-Z0-9_-]+)", url)
        if match:
            return f"https://www.youtube.com/watch?v={match.group(1)}"
    elif "youtu.be/" in url:
        match = re.search(r"youtu\.be/([a-zA-Z0-9_-]+)", url)
        if match:
            return f"https://www.youtube.com/watch?v={match.group(1)}"
    return url


def is_youtube_url(url: str) -> bool:
    """Check if a URL is a valid YouTube URL."""
    return any(re.search(pattern, url) for pattern in YOUTUBE_PATTERNS)


def extract_video_id(url: str) -> str | None:
    """Extract video ID from YouTube URL."""
    match = re.search(r"v=([a-zA-Z0-9_-]+)", url)
    if match:
        return match.group(1)

    match = re.search(r"youtu\.be/([a-zA-Z0-9_-]+)", url)
    if match:
        return match.group(1)

    return None

File: llm-harness-js/src/utils/youtubeUtils.ts
const YOUTUBE_PATTERNS = [
  /youtube\.com\/watch\?v=/,
  /youtu\.be\//,
  /youtube\.com\/embed\//,
  /youtube\.com\/v\//,
];

export function cleanText(text: string): string {
  return text
    .replace(/\n{3,}/g, "\n\n")
    .replace(/ {2,}/g, " ")
    .trim();
}

export function cleanYoutubeUrl(url: string): string {
  if (url.includes("youtube.com/watch")) {
    const match = /v=([a-zA-Z0-9_-]+)/.exec(url);
    if (match?.[1]) {
      return `https://www.youtube.com/watch?v=${match[1]}`;
    }
  } else if (url.includes("youtu.be/")) {
    const match = /youtu\.be\/([a-zA-Z0-9_-]+)/.exec(url);
    if (match?.[1]) {
      return `https://www.youtube.com/watch?v=${match[1]}`;
    }
  }
  return url;
}

export function isYoutubeUrl(url: string): boolean {
  return YOUTUBE_PATTERNS.some((pattern) => pattern.test(url));
}

export function extractVideoId(url: string): string | null {
  const longMatch = /v=([a-zA-Z0-9_-]+)/.exec(url);
  if (longMatch?.[1]) {
    return longMatch[1];
  }

  const shortMatch = /youtu\.be\/([a-zA-Z0-9_-]+)/.exec(url);
  if (shortMatch?.[1]) {
    return shortMatch[1];
  }

  return null;
}
